scylla-cdc/src/stream_generations.rs

```diff
@@ -0,0 +1,407 @@
+use futures::stream::StreamExt;
+use scylla::frame::response::result::Row;
+use scylla::frame::value::Timestamp;
+use scylla::{IntoTypedRows, Session};
+use std::error::Error;
+use std::sync::Arc;
+use uuid::Uuid;
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub struct Generation(chrono::Duration);
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub struct StreamID(Vec<u8>);
+
+// Component responsible for managing stream generations.
+pub struct GenerationFetcher {
+    generations_table_name: String,
+    streams_table_name: String,
+    session: Arc<Session>,
+}
+
+impl GenerationFetcher {
+    // Creates new instance of GenerationFetcher.
+    pub fn new(session: &Arc<Session>) -> GenerationFetcher {
+        GenerationFetcher {
+            generations_table_name: "system_distributed.cdc_generation_timestamps".to_string(),
+            streams_table_name: "system_distributed.cdc_streams_descriptions_v2".to_string(),
+            session: Arc::clone(session),
+        }
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_all_stream_generations_query(&self) -> String {
+        format!(
+            r#"
+    SELECT time
+    FROM {}
+    WHERE key = 'timestamps';
+    "#,
+            self.generations_table_name
+        )
+    }
+
+    // Internal function configurable for testing purposes.
+    pub async fn fetch_all_generations(&self) -> Result<Vec<Generation>, Box<dyn Error>> {
+        let mut rows = self
+            .session
+            .query_iter(self.get_all_stream_generations_query(), &[])
+            .await?
+            .into_typed::<(chrono::Duration,)>();
+
+        let mut generations = vec![];
+
+        while let Some(next_row) = rows.next().await {
+            let (generation,) = next_row?;
+            generations.push(Generation(generation));
+        }
+
+        Ok(generations)
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_generation_by_timeuuid_query(&self) -> String {
+        format!(
+            r#"
+    SELECT time
+    FROM {}
+    WHERE key = 'timestamps'
+    AND time <= dateOf( ? )
+    LIMIT 1;
+    "#,
+            self.generations_table_name
+        )
+    }
+
+    // Internal function configurable for testing purposes.
+    pub async fn fetch_generation_by_timeuuid(
+        &self,
+        time: &Uuid,
+    ) -> Result<Option<Generation>, Box<dyn Error>> {
```

Comment:
Does this really cover the requirement described as 

> For a given timestamp, which generation contains it

? I'm not sure if a timestamp is the same thing as an Uuid. Also, it kinda makes my work in #4 unfeasible.
@piodul what do you think?

Reply:
A timestamp is not the same as a timeuuid, but a timeuuid contains a timestamp + some other parts. This function does not use "the other parts" (because `dateOf` gets rid of them), and it's easier to convert a timeuuid to a timestamp, so I agree that this function should receive a timestamp, not a timeuuid.